CRM/Core/Form/Search.php

```diff
@@ -123,6 +144,60 @@ public function buildQuickform() {
     $this->addTaskMenu($tasks);
   }
 
+  /**
+   * Add any fields described in metadata to the form.
+   *
+   * The goal is to describe all fields in metadata and handle from metadata rather
+   * than existing ad hoc handling.
+   */
+  public function addFormFieldsFromMetadata() {
+    $this->_action = CRM_Core_Action::ADVANCED;
+    foreach ($this->getSearchFieldMetadata() as $entity => $fields) {
+      foreach ($fields as $fieldName => $fieldSpec) {
+        $this->addField($fieldName, ['entity' => $entity]);
+      }
+    }
+  }
+
+  /**
+   * Get the validation rule to apply to a function.
+   *
+   * Alphanumeric is designed to always be safe & for now we just return
+   * that but in future we can use tighter rules for types like int, bool etc.
+   *
+   * @param string $entity
+   * @param string $fieldName
+   *
+   * @return string
+   */
+  protected function getValidationTypeForField($entity, $fieldName) {
+    switch ($this->getSearchFieldMetadata()[$entity][$fieldName]['type']) {
+      case CRM_Utils_Type::T_BOOLEAN:
+        return 'Boolean';
+
+      default:
+        return 'Alphanumeric';
+    }
```

Comment:
How about ?: 
```
protected function getValidationTypeForField($entity, $fieldName) {
 $type = CRM_Utils_Type::typeToString($this->getSearchFieldMetadata()[$entity][$fieldName]['type']);
 return  $type !== '' ? $type : 'Aplhanumeric';
}
```

Reply:
@monishdeb I agree that's more concise - the reason I did the switch is that I expected it to be filled out over time with different types

Reply:
Oh ok, agree.

Reply:
@eileenmcnaughton as I suspected, I am getting an warning for this check as:
![screen shot 2018-12-03 at 12 09 43 pm](https://user-images.githubusercontent.com/3735621/49357192-7dedf500-f6f4-11e8-9ef7-84669916fc99.png)

Reply:
@monishdeb OK - I changed it to ignore if $_POST is set - also rebased